clarity/src/vm/callables.rs

```diff
@@ -300,3 +339,216 @@ impl FunctionIdentifier {
         }
     }
 }
+
+// Implicitly cast principals to traits and traits to other traits as needed,
+// recursing into compound types. This function does not check for legality of
+// these casts, as that is done in the type-checker. Note: depth of recursion
+// should be capped by earlier checks on the types/values.
+fn implicit_cast(type_sig: &TypeSignature, value: &Value) -> Result<Value> {
+    Ok(match (type_sig, value) {
+        (
+            TypeSignature::OptionalType(inner_type),
+            Value::Optional(OptionalData {
+                data: Some(inner_value),
+            }),
+        ) => Value::Optional(OptionalData {
+            data: Some(Box::new(implicit_cast(inner_type, inner_value)?)),
+        }),
+        (
+            TypeSignature::ResponseType(inner_types),
+            Value::Response(ResponseData { committed, data }),
+        ) => Value::Response(ResponseData {
+            committed: *committed,
+            data: Box::new(implicit_cast(
+                if *committed {
+                    &inner_types.0
+                } else {
+                    &inner_types.1
+                },
+                data,
+            )?),
+        }),
+        (
+            TypeSignature::SequenceType(SequenceSubtype::ListType(list_type)),
+            Value::Sequence(SequenceData::List(ListData {
+                data,
+                type_signature,
+            })),
+        ) => {
+            let mut values = Vec::new();
```

Comment:
Minor performance tweak: you can use `Vec::with_capacity()` to avoid any additional allocations when pushing data to `values`.

Reply:
Updated in 64fe8486c